File: test/language/service_types/cpp/SimpleServiceTest.cpp
#include <fstream>
#include <memory>

#include "gtest/gtest.h"

#include "zserio/ServiceException.h"
#include "service_types/simple_service/SimpleService.h"

using namespace zserio::literals;

namespace service_types
{
namespace simple_service
{

using allocator_type = SimpleService::Client::allocator_type;

namespace
{
    struct FakeContext
    {
        bool seenByService = false;
    };
}

class SimpleServiceImpl : public SimpleService::Service
{
public:
    Response powerOfTwoImpl(const Request& request, void* context) override
    {
        if (context != nullptr)
            static_cast<FakeContext*>(context)->seenByService = true;

        int32_t value = request.getValue();
        if (value < 0)
            value = -value;
        return Response(static_cast<uint64_t>(value) * static_cast<uint64_t>(value), get_allocator_ref());
    }
};

class SimpleServiceTest : public ::testing::Test
{
public:
    SimpleServiceTest()
    :   client(service),
        movedService(createService()),
        movedClient(createClient())
    {}

private:
    // check that move constructors work correctly
    SimpleServiceImpl createService()
    {
        // cannot be const because we need to move-construct it
        SimpleServiceImpl createdService;
        return createdService;
    }

    SimpleService::Client createClient()
    {
        // cannot be const because we need to move-construct it
        SimpleService::Client createdClient(movedService);
        return createdClient;
    }

protected:
    SimpleServiceImpl service;
    SimpleService::Client client;
    SimpleServiceImpl movedService;
    SimpleService::Client movedClient;
};

TEST_F(SimpleServiceTest, serviceFullName)
{
    ASSERT_EQ("service_types.simple_service.SimpleService"_sv,
            SimpleService::Service::serviceFullName());
}

TEST_F(SimpleServiceTest, methodNames)
{
    ASSERT_EQ("powerOfTwo"_sv, SimpleService::Service::methodNames()[0]);
}

TEST_F(SimpleServiceTest, powerOfTwo)
{
    Request request;

    request.setValue(13);
    Response response = client.powerOfTwoMethod(request);
    ASSERT_EQ(169, response.getValue());
    request.setValue(-13);
    response = movedClient.powerOfTwoMethod(request);
    ASSERT_EQ(169, response.getValue());
    request.setValue(2);
    response = client.powerOfTwoMethod(request);
    ASSERT_EQ(4, response.getValue());
    request.setValue(-2);
    response = movedClient.powerOfTwoMethod(request);
    ASSERT_EQ(4, response.getValue());
}

TEST_F(SimpleServiceTest, invalidServiceMethod)
{
    zserio::BlobBuffer<allocator_type> responseData;
    ASSERT_THROW(service.callMethod("nonexistentMethod"_sv, {}, responseData), zserio::ServiceException);
}

TEST_F(SimpleServiceTest, callWithContext)
{
    FakeContext fakeContext;
    ASSERT_FALSE(fakeContext.seenByService);
    Request request{10};
    Response response = client.powerOfTwoMethod(request, &fakeContext);
    ASSERT_EQ(100, response.getValue());
    ASSERT_TRUE(fakeContext.seenByService);

    fakeContext.seenByService = false;
    response = movedClient.powerOfTwoMethod(request, &fakeContext);
    ASSERT_EQ(100, response.getValue());
    ASSERT_TRUE(fakeContext.seenByService);
}

} // namespace simple_service
} // namespace service_types

File: test/language/service_types/python/SimpleServiceTest.py
import zserio

import ServiceTypes

class FakeContext:
    def __init__(self):
        self.seenByService = False

class LocalServiceClient(zserio.ServiceClientInterface):
    def __init__(self, service):
        self._service = service

    def call_method(self, method_name, request, context = None):
        response = self._service.call_method(method_name, request.byte_array, context)

        return response.byte_array

class SimpleServiceTest(ServiceTypes.TestCase):
    @classmethod
    def setUpClass(cls):
        super(SimpleServiceTest, cls).setUpClass()

        class Service(cls.api.SimpleService.Service):
            def __init__(self, api):
                super().__init__()
                self.api = api

            @staticmethod
            def _power_of_two_impl(request, context):
                if context:
                    context.seenByService = True

                value = request.value
                response = cls.api.Response(value**2)
                return response

            def _power_of_two_raw_impl(self, request_data, context):
                request = zserio.deserialize_from_bytes(self.api.Request, request_data)
                response = SimpleServiceTest.Service._power_of_two_impl(request, context)
                return zserio.serialize_to_bytes(response)

        cls.Service = Service

    def setUp(self):
        self.service = self.Service(self.api)
        self.client = self.api.SimpleService.Client(LocalServiceClient(self.service))

    def testServiceFullName(self):
        self.assertEqual("service_types.simple_service.SimpleService", self.service.service_full_name)

    def testMethodNames(self):
        self.assertEqual("powerOfTwo", self.service.method_names[0])

    def testPowerOfTwo(self):
        request = self.api.Request(13)
        self.assertEqual(169, self.client.power_of_two(request).value)
        request.value = -13
        self.assertEqual(169, self.client.power_of_two(request).value)
        request.value = 2
        self.assertEqual(4, self.client.power_of_two(request).value)
        request.value = -2
        self.assertEqual(4, self.client.power_of_two(request).value)

    def testPowerOfTwoRaw(self):
        request = self.api.Request(13)
        request_data = zserio.serialize_to_bytes(request)
        response_data = self.client.power_of_two_raw(request_data)
        response = zserio.deserialize_from_bytes(self.api.Response, response_data)
        self.assertEqual(169, response.value)

    def testInvalidServiceMethod(self):
        with self.assertRaises(zserio.ServiceException):
            self.service.call_method("nonexistentMethod", bytes())

    def testCallWithContext(self):
        fakeContext = FakeContext()
        self.assertFalse(fakeContext.seenByService)
        request = self.api.Request(10)
        response = self.client.power_of_two(request, fakeContext)
        self.assertEqual(100, response.value)
        self.assertTrue(fakeContext.seenByService)
